walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,107 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.List;
+
 
 public final class FindMeetingQuery {
+
+  /*
+   Criterion functional interface is used to define who's (mandatory or all attendees')
+   occupied hours should be taken in account in findAvailableTimeRanges function
+   */
+  @FunctionalInterface
+  private interface Criterion {
+    boolean satisfies(int x);
+  }
+  private final Criterion COMMON_CRITERION = x -> x == 0;
+  private final Criterion MANDATORY_CRITERION = x -> x <= 1;
+
+  private static final int DURATION_DAY = 60 * 24;
+
+  private int getStart(Event event) {
+    return event.getWhen().start();
+  }
+
+  private int getEnd(Event event) {
+    return event.getWhen().end();
+  }
+
+  private boolean haveCommonAttendees(Collection<String> a, Collection<String> b) {
+    return new HashSet<>(a).removeAll(b);
+  }
+
+  /*
+  The algorithm skips all minutes that don't meet the criterion,
+  memorises the index of the first fitting minute in a row and goes until
+  it finds the new unfitting minute. Adds the range to the list if
+  if it's not less than the target duration.
```

Comment:
```suggestion
  it's not less than the target duration.
```
For your consideration: Try to write comments to describe the desired outcome instead of translating code directly into prose. "Find all time ranges where all considered attendees have sufficient time in their schedule at the same time." - As very short example.